### PR TITLE
HW7

### PR DESCRIPTION
**Pull request by Ragnoji**
### Summary

A brief description of the pull request.

### How it works

Screenshots/screencasts of the pull request introduced functionality.

### Test plan

List of steps to manually test introduced functionality:

- Log in
- Go to http://localhost:3000

### Deploy notes

Notes regarding deployment the contained body of work.
These should note any db migrations, services, scripts, etc.

### Checklist:

- [x] I have performed a self-review of my own code
- [x] My pull request is shorter than 500 lines

### CODE REVIEW

File: app/interactors/comments/update.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+module Comments
+  class Update
+    include Interactor
+
+    delegate :comment, :comment_params, to: :context
+
+    def call
+      context.fail(error: "Invalid Data") unless comment.update(comment_params)
```
Comment (RasimKhusaenov) [inline app/interactors/comments/update.rb:8]: Use bang method

File: app/interactors/projects/update_organize.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+module Projects
+  class UpdateOrganize
```
Comment (RasimKhusaenov) [inline app/interactors/projects/update_organize.rb:2]: Update

File: app/interactors/projects/update_organize.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    include Interactor::Organizer
+
+    organize Projects::Update,
```
Comment (RasimKhusaenov) [inline app/interactors/projects/update_organize.rb:5]: You can name it like `Projects::Update::Execute`

File: app/interactors/tasks/create.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+module Tasks
+  class Create
+    include Interactor::Organizer
+
+    organize Tasks::PrepareParams,
+             Tasks::Save, Tasks::Create::SendNotification
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/create.rb:6]: Write one interactor per line please

File: app/mailers/task_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
+class TaskMailer < ApplicationMailer
+  def task_created_to_members(project, task)
+    @project = project
+    @task = task
+
+    mail(to: ProjectMembership.where(role: :member, project: project).map { |p| p.user.email })
```
Comment (RasimKhusaenov) [inline app/mailers/task_mailer.rb:6]: Can you use `has_many` association pros? `project.users.where(project_memberships: { role: :member }).pluck(:email)`

File: app/mailers/task_mailer.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    mail(to: ProjectMembership.find_by(role: :owner, project: project).user.email)
+  end
+
+  def task_destroyed_to_members(project, task, user)
+    @project = project
+    @task = task
+    @user = user
+
+    recipients = ProjectMembership.where(role: :member, project: project)
+                                  .filter_map { |p| p.user.email unless p.user.eql? user }
```
Comment (RasimKhusaenov) [inline app/mailers/task_mailer.rb:44]: `project.users.where(project_memberships: { role: :member }).where.not(id: user).pluck(:email)`

File: app/mailers/task_mailer.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    @project = project
+    @task = task
+
+    mail(to: user.email)
+  end
+
+  def task_destroyed_to_owner(project, task, user)
+    @project = project
+    @task = task
+    @user = user
```
Comment (RasimKhusaenov) [inline app/mailers/task_mailer.rb:33]: Why you need it?

File: app/interactors/tasks/destroy/send_notification.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
+module Tasks
+  class Destroy
+    class SendNotification
+      include Interactor
+
+      delegate :project, :task, :user, to: :context
+
+      def call
+        mail_owner.deliver_later unless ProjectMembership.find_by(project: project, user: user).owner?
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/destroy/send_notification.rb:9]: If I am a project member, will I receive 1 email or 2?

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/interactors/tasks/destroy/send_notification.rb (hunk 8 above)
Comment (Ragnoji) [reply app/interactors/tasks/destroy/send_notification.rb:9]: No, if you were project member, this part of code would send email only to project owner:

```
def mail_owner
  @mail_owner ||= TaskMailer.task_destroyed_to_owner(project, task)
end
...
class TaskMailer < ApplicationMailer
  def task_destroyed_to_owner(project, task)
    mail(to: project.users.where(project_memberships: { role: :owner }).pluck(:email))
  end
end
```

And mail_members would send mail to every member of project except owner and you(initiator):
```
def mail_members
  @mail_members ||= TaskMailer.task_destroyed_to_members(project, task, user)
end
...
class TaskMailer < ApplicationMailer
  def task_destroyed_to_members(project, task, user)
    recipients = project.users.where(project_memberships: { role: :member }).where.not(id: user).pluck(:email)
  
    mail(to: recipients)
  end
end
```

File: app/controllers/tasks_controller.rb [hunk 9]
```diff
@@ -5,22 +5,20 @@ class TasksController < ApplicationController
   before_action -> { authorize! @task }, only: %i[show edit update destroy]
 
   def index
-    @task = Task.new(project: @project)
-    authorize! @task
+    authorize! Task.new(project: @project)
```
Comment (RasimKhusaenov) [inline app/controllers/tasks_controller.rb:8]: How about move it to `before_action`?

File: app/interactors/tasks/create/send_notification.rb [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+      end
+
+      private
+
+      def mail_owner
+        @mail_owner ||= TaskMailer.task_created_to_owner(project, task)
+      end
+
+      def mail_members
+        @mail_members ||= TaskMailer.task_created_to_members(project, task)
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/create/send_notification.rb:20]: I think additional methods for these mails are unnecessary. You can put them in the `call` method.

File: app/interactors/tasks/destroy/send_notification.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      def mail_owner
+        @mail_owner ||= TaskMailer.task_destroyed_to_owner(project, task)
+      end
+
+      def mail_members
+        @mail_members ||= TaskMailer.task_destroyed_to_members(project, task, user)
+      end
+
+      def mail_initiator
+        @mail_initiator ||= TaskMailer.task_destroyed_to_initiator(project, task, user)
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/destroy/send_notification.rb:25]: Why you memoize them?

File: app/interactors/tasks/prepare_params.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    delegate :task_params, :project, to: :context
+
+    def call
+      context.prepared_params = prepared_params
+    end
+
+    private
+
+    def prepared_params
+      @prepared_params ||= task_params.merge({ project: project })
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/prepare_params.rb:14]: Why you memoize it?

File: app/interactors/tasks/update/send_notification.rb [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+      delegate :project, :task, to: :context
+
+      def call
+        mail_members.deliver_later
+      end
+
+      private
+
+      def mail_members
+        @mail_members ||= TaskMailer.task_updated(project, task)
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/update/send_notification.rb:15]: Why you memoize it?

File: app/interactors/tasks/update.rb [hunk 14]
```diff
@@ -0,0 +1,13 @@
+module Tasks
+  class Update
+    class Execute
+      include Interactor
+
+      delegate :task, :task_params, to: :context
+
+      def call
+        context.fail(error: "Invalid Data") unless task.update(task_params)
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/update.rb:9]:
```suggestion
        context.fail!(error: "Invalid Data") unless task.update(task_params)
```

File: app/mailers/project_mailer.rb [hunk 15]
```diff
@@ -4,4 +4,16 @@ def project_created(project, user)
 
     mail(to: user.email)
   end
+
+  def project_updated(project)
+    @project = project
+
+    mail(to: project.users.map(&:email))
```
Comment (RasimKhusaenov) [inline app/mailers/project_mailer.rb:11]: It is better to use `pluck`

File: app/interactors/projects/destroy.rb [hunk 16]
```diff
@@ -0,0 +1,13 @@
+module Projects
+  class Destroy
+    class Execute
+      include Interactor
+
+      delegate :project, to: :context
+
+      def call
+        context.fail(error: "Invalid Data") unless project.destroy
```
Comment (RasimKhusaenov) [inline app/interactors/projects/destroy.rb:9]:
```suggestion
        context.fail!(error: "Invalid Data") unless project.destroy
```

File: app/interactors/tasks/destroy/send_notification.rb [hunk 17]
```diff
@@ -0,0 +1,21 @@
+module Tasks
+  class Destroy
+    class SendNotification
+      include Interactor
+
+      delegate :project, :task, :user, to: :context
+
+      def call
+        mail_owner.deliver_later unless ProjectMembership.find_by(project: project, user: user).owner?
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/destroy/send_notification.rb:9]: Maybe if owner, not unless?

File: app/interactors/tasks/update.rb [hunk 18]
```diff
@@ -0,0 +1,13 @@
+module Tasks
+  class Update
+    class Execute
+      include Interactor
+
+      delegate :task, :task_params, to: :context
+
+      def call
+        context!.fail(error: "Invalid Data") unless task.update(task_params)
```
Comment (RasimKhusaenov) [inline app/interactors/tasks/update.rb:9]: Misspell

File: app/interactors/tasks/destroy/send_notification.rb (hunk 17 above)
Comment (Ragnoji) [reply app/interactors/tasks/destroy/send_notification.rb:9]: it handles logic of sending message to owner that member of his project deleted task so unless should be used to check if user isnt the owner himself, i could've done this check inside mailer but i chose not to

File: app/interactors/tasks/update.rb (hunk 18 above)
Comment (Ragnoji) [reply app/interactors/tasks/update.rb:9]:
```suggestion
        context.fail!(error: "Invalid Data") unless task.update(task_params)
```

File: app/interactors/tasks/destroy/send_notification.rb (hunk 17 above)
Comment (RasimKhusaenov) [reply app/interactors/tasks/destroy/send_notification.rb:9]: Then you can move checking user role to method "initiator_is_owner?"

**Review by RasimKhusaenov — APPROVED** [review]

**Pull request merged**